build_tools/analyze_memory_logs.py

```diff
@@ -0,0 +1,336 @@
+                )
+
+        lines.append("")
+        lines.append("=" * 80)
+
+        return "\n".join(lines)
+
+    def write_github_summary(self, phase_analyses: List[Dict[str, Any]]):
+        """Write analysis to GitHub Actions step summary."""
+        if "GITHUB_STEP_SUMMARY" not in os.environ:
+            return
+
+        try:
+            with open(os.environ["GITHUB_STEP_SUMMARY"], "a") as f:
+                f.write("\n## Memory Usage Analysis\n\n")
+                f.write("### Peak Memory Usage by Phase\n\n")
+                f.write("| Phase | Peak Memory | Avg Memory | Peak Swap | Severity |\n")
+                f.write("|-------|-------------|------------|-----------|----------|\n")
+
+                for analysis in phase_analyses[:10]:  # Top 10
+                    prefix = {
+                        "CRITICAL": ":red_circle:",
+                        "HIGH": ":orange_circle:",
+                        "MEDIUM": ":yellow_circle:",
+                        "LOW": ":green_circle:",
+                    }.get(analysis["severity"], ":white_circle:")
+
+                    f.write(
+                        f"| {analysis['phase']} | "
+                        f"{analysis['max_memory_percent']:.1f}% | "
+                        f"{analysis['avg_memory_percent']:.1f}% | "
+                        f"{analysis['max_swap_percent']:.1f}% | "
+                        f"{prefix} {analysis['severity']} |\n"
+                    )
+
+                # Add warnings for critical phases
+                critical = [
+                    a for a in phase_analyses if a["severity"] in ["CRITICAL", "HIGH"]
+                ]
+                if critical:
+                    f.write("\n### :warning: Phases with High Memory Usage\n\n")
+                    for analysis in critical:
+                        f.write(
+                            f"- **{analysis['phase']}**: {analysis['max_memory_percent']:.1f}% peak memory\n"
+                        )
+
+                f.write("\n")
+        except Exception as e:
+            print(f"Warning: Failed to write GitHub summary: {e}", file=sys.stderr)
+
+
+def main():
+    parser = argparse.ArgumentParser(
+        description="Analyze memory logs from CI builds",
+        formatter_class=argparse.RawDescriptionHelpFormatter,
+    )
+
+    parser.add_argument(
+        "--log-dir",
+        type=Path,
+        default=Path("build/logs"),
+        help="Directory containing memory log files (default: build/logs)",
+    )
+
+    parser.add_argument(
+        "--detailed",
+        action="store_true",
+        help="Generate detailed report with per-phase breakdowns",
+    )
+
+    parser.add_argument(
+        "--output", type=Path, help="Write report to file instead of stdout"
+    )
+
+    parser.add_argument(
+        "--github-summary",
+        action="store_true",
+        help="Write summary to GITHUB_STEP_SUMMARY",
+    )
+
+    args = parser.parse_args()
```

Comment:
Other scripts in this project forward `argv` to `main()` functions explicitly, to allow for other scripts (like unit tests) to call that function with custom arguments